snuba/web/query.py

```diff
@@ -195,48 +224,87 @@ def raw_query(
     return result
 
 
-def log_query_and_update_stats(
+def update_query_list_and_stats(
     request: Request,
     sql: str,
     timer: Timer,
     stats: MutableMapping[str, Any],
-    status: str,
+    query_list: MutableSequence[ClickhouseQueryMetadata],
     query_settings: Mapping[str, Any],
+    trace_id: str,
+    status: str,
 ) -> MutableMapping:
     """
     If query logging is enabled then logs details about the query and its status, as
     well as timing information.
     Also updates stats with any relevant information and returns the updated dict.
     """
-
     stats.update(query_settings)
+
+    query_list.append(
+        ClickhouseQueryMetadata(
+            request=request.body,
+            sql=sql,
+            timer=timer,
+            stats=stats,
+            status=status,
+            trace_id=trace_id,
+        )
+    )
+
+    return stats
+
+
+def record_query(request: Request, timer: Timer, query_list: MutableSequence[ClickhouseQueryMetadata]) -> None:
     if settings.RECORD_QUERIES:
+        last_query = query_list[-1]
         # send to redis
         state.record_query(
             {
                 "request": request.body,
-                "sql": sql,
-                "timing": timer,
-                "stats": stats,
-                "status": status,
+                "referrer": http_request.referrer,
+                "timing": last_query.timer.for_json(),
+                "status": last_query.status,
+                "query_list": [q.to_dict() for q in query_list],
             }
         )
 
         timer.send_metrics_to(
             metrics,
             tags={
-                "status": str(status),
-                "referrer": stats.get("referrer", "none"),
-                "final": str(stats.get("final", False)),
+                "status": str(last_query.status),
+                "referrer": last_query.stats.get("referrer", "none"),
+                "final": str(last_query.stats.get("final", False)),
             },
-            mark_tags={"final": str(stats.get("final", False))},
+            mark_tags={"final": str(last_query.stats.get("final", False))},
```

Comment:
The fact that you need to rely on the last query you ran to retrieve some fields that are proper of the Snuba request and not of the Clickhouse query (referrer and  status) hints that this may not be the right abstraction, and it would actually break if we needed to record in the stats something about a query that failed during parsing (which would make sense). 

I think the abstraction we are missing here is a SnubaQueryMetadata dataclass that you build at the beginning of parse_and_run with the Snuba query specific fields, and you fill in multiple times with ClickhouseQueryMetadata objects. What do you think about this ?

Reply:
Yeah, I like this idea a lot